### PR TITLE
Add 'manifests' package that contains immutable input and build manifest classes

### PR DESCRIPTION
**Pull request by camerski**
### Description
Provides `InputManifest` and `BuildManifest` classes in the `manifests` package.
 
### Issues Resolved
https://github.com/opensearch-project/opensearch-build/issues/157
 
### Check List
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: bundle-workflow/python/manifests/__init__.py
Comment (dblock) [inline bundle-workflow/python/manifests/__init__.py:1]: If you're going to write a sentence, starting with a capital T, finish it with a period :)

File: bundle-workflow/python/manifests/build_manifest.py
Comment (dblock) [inline bundle-workflow/python/manifests/build_manifest.py:20]: Could we make this `commit_id` or `ref`? Otherwise we'll end up with things like `buildmanifestthatiscool`.

File: bundle-workflow/python/manifests/build_manifest.py
Comment (mch2) [inline bundle-workflow/python/manifests/build_manifest.py:15]: We should include platform/architecture here as well

**Review by dblock — COMMENTED** [review]
- I think you also need a base `Manifest` class and sink common logic such as version into it. 
- I don't think we should merge this until the classes are used, otherwise it's not really a refactor.
- plus minor stuff.

File: bundle-workflow/python/manifests/build_manifest.py
Comment (camerski) [reply bundle-workflow/python/manifests/build_manifest.py:15]: Yes, we should.

**Review by mch2 — APPROVED** [review]
lgtm. with minor comment

**Comment by camerski** [comment]
> * I think you also need a base `Manifest` class and sink common logic such as version into it.

Inheritance is not the right approach here, composition is. We could possibly extract some parts to compose, but it's not worth it at this point.

> * I don't think we should merge this until the classes are used, otherwise it's not really a refactor.

Merging dead code is totally OK so that the reviewers of the upcoming refactors can see how they fit together. If you feel strongly then I can hold off on the merge until the rest of the PRs are approved, but I don't think that is necessary.

File: bundle-workflow/python/manifests/build_manifest.py
Comment (camerski) [reply bundle-workflow/python/manifests/build_manifest.py:20]: `commitid` and `ref` are capturing different things. We only really *need* the `commitid`, but `ref` is helpful for humans to quickly know if this was a build from `main` or `1.0` or whatever branch it was.

I don't quite understand what you mean by `buildmanifestthatiscool`...can you explain?

File: bundle-workflow/python/manifests/__init__.py
Comment (camerski) [reply bundle-workflow/python/manifests/__init__.py:1]: I grammar the way I wants

File: bundle-workflow/python/manifests/build_manifest.py
Comment (mch2) [reply bundle-workflow/python/manifests/build_manifest.py:20]: I'd like to be able to feed BuildManifests back into the build system and rebuild components from the same sha.  Use case:  After the initial snapshot build we want to rebuild without a snapshot tag from the same sha's and promote to staging.  If we are using 'ref' from InputManifest, I'd rather use 'ref' and make guarantees that in a BuildManifest this will always be a sha

**Comment by dblock** [comment]
> Merging dead code is totally OK so that the reviewers of the upcoming refactors can see how they fit together. If you feel strongly then I can hold off on the merge until the rest of the PRs are approved, but I don't think that is necessary.

It'd be fine as dead code. But I feel strongly because this is dead code that duplicates existing logic, making the codebase net worse. Finish the refactor of replacing the manifests in https://github.com/opensearch-project/opensearch-build/tree/main/bundle-workflow/python/lib with this code, and then we can merge?

**Comment by camerski** [comment]
> > Merging dead code is totally OK so that the reviewers of the upcoming refactors can see how they fit together. If you feel strongly then I can hold off on the merge until the rest of the PRs are approved, but I don't think that is necessary.
> 
> It'd be fine as dead code. But I feel strongly because this is dead code that duplicates existing logic, making the codebase net worse. Finish the refactor of replacing the manifests in https://github.com/opensearch-project/opensearch-build/tree/main/bundle-workflow/python/lib with this code, and then we can merge?

Sure, that's fine. The rest of the refactoring is coming up this afternoon. I've done all the code changes; I'm just picking them apart into standalone PRs so that the reviews are easy.

File: bundle-workflow/python/manifests/build_manifest.py
Comment (dblock) [reply bundle-workflow/python/manifests/build_manifest.py:21]: I am just saying that `commitid` should be `commit_id`, underscore, because it's not 1 word.

A "commit ID" is a SHA-1 of all the changes, see https://stackoverflow.com/questions/29106996/what-is-a-git-commit-id. A "ref" is an indirect reference to a commit. The most direct way to reference a commit is via its SHA-1 hash, but it can be a branch, tag, etc.

So I think @mch2 you're saying all the right things and you want a "commit ID", and @camerski is right here to call that a commit ID and not a "ref".

**Comment by camerski** [comment]
> > > Merging dead code is totally OK so that the reviewers of the upcoming refactors can see how they fit together. If you feel strongly then I can hold off on the merge until the rest of the PRs are approved, but I don't think that is necessary.
> > 
> > 
> > It'd be fine as dead code. But I feel strongly because this is dead code that duplicates existing logic, making the codebase net worse. Finish the refactor of replacing the manifests in https://github.com/opensearch-project/opensearch-build/tree/main/bundle-workflow/python/lib with this code, and then we can merge?
> 
> Sure, that's fine. The rest of the refactoring is coming up this afternoon. I've done all the code changes; I'm just picking them apart into standalone PRs so that the reviews are easy.

@dblock Assuming you're OK with that plan, can I have an approval? I obviously won't execute the merge until the other changes are approved, but I don't want to leave this change hanging.

**Comment by dblock** [comment]
> @dblock Assuming you're OK with that plan, can I have an approval? I obviously won't execute the merge until the other changes are approved, but I don't want to leave this change hanging.

Code here looks OK to me. Let's see those other changes on top!

File: bundle-workflow/python/manifests/build_manifest.py
Comment (camerski) [reply bundle-workflow/python/manifests/build_manifest.py:21]: For that use case I would generate a new InputManifest from the BuildManifest, rather than reading a BuildManifest and trying to treat it as an InputManifest. Does that make sense?

**Comment by dblock** [comment]
You should merge https://github.com/opensearch-project/opensearch-build/pull/162 first though because otherwise @mch2 will have to go back and redo that work on top of your upcoming changes.

File: bundle-workflow/python/manifests/build_manifest.py
Comment (mch2) [reply bundle-workflow/python/manifests/build_manifest.py:21]: Makes sense - I think a better approach for my use case is to create a separate promotion script that creates a new InputManifest using the commit_id in a built BundleManifest as a component's ref, and then feed it back to the build system.

**Comment by camerski** [comment]
> You should merge #162 first though because otherwise @mch2 will have to go back and redo that work on top of your upcoming changes.

Discussed with @mch2 and agreed that this change can go first. I'll rebase #162 on top of this one.

**Comment by camerski** [comment]
> > @dblock Assuming you're OK with that plan, can I have an approval? I obviously won't execute the merge until the other changes are approved, but I don't want to leave this change hanging.
> 
> Code here looks OK to me minus the naming of `commit_id`. Let's see those other changes on top!

@dblock Renamed.

File: bundle-workflow/python/manifests/build_manifest.py
Comment (gaiksaya) [reply bundle-workflow/python/manifests/build_manifest.py:15]: Should we add platform as well? linux/mac/windows?

Comment (camerski) [reply bundle-workflow/python/manifests/build_manifest.py:15]: Graah, yeah, platform is not the same as architecture. My bad :)

I have a few changes stacked on top of this...can I add platform as a new PR?

Comment (gaiksaya) [reply bundle-workflow/python/manifests/build_manifest.py:15]: Sure if you remember it! :)

**Review by gaiksaya — APPROVED** [review]
To do: Add platform in new PR

**Comment by camerski** [comment]
Replaced with https://github.com/opensearch-project/opensearch-build/pull/170

**Pull request closed without merging**
